common/elasticsearch/bulk/bulk_delete_request.go

```diff
@@ -0,0 +1,142 @@
+// The MIT License (MIT)
+
+// Copyright (c) 2017-2020 Uber Technologies Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package bulk
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+)
+
+// -- Bulk delete request --
+
+// BulkDeleteRequest is a request to remove a document from Elasticsearch.
+//
+// See https://www.elastic.co/guide/en/elasticsearch/reference/7.0/docs-bulk.html
+// for details.
+type BulkDeleteRequest struct {
+	GenericBulkableRequest
```

Comment:
after a bunch of reading: I think there are a few issues with these three special request types.

Special types in general: no complaint at all!  More types is more good IMO.  But for these:

- this is declaring a *struct field* which contains a generic bulkable request, but it doesn't seem to be assigned anywhere.
  - I'm guessing this is intended as a "this struct implements this interface", but that's not what this does; if anything it makes it _riskier_ since nothing is ever assigned to it.  it's likely null -> the BulkDeleteRequest still implements these methods because it's embedded, but now you have no warning if you forget a method, it'll just panic.
  - try a `var _ GenericBulkableRequest = (*BulkDeleteRequest)(nil)` as a type assertion if the goal is "must implement X interface".
- these are in `bulk`, but only `os2` uses them.
  - is the plan to migrate all implementations to them?  that seems fine (yay types rather than enum + generic!), it's just not documented for future readers prior to that being built.  it doesn't have to be in this PR, but there should probably be something ready-to-go immediately after, or comments so it's not confusing in the future if that gets pushed back.

Reply:
re this `bulk` package:
There's no plan to migrate old versions (v6,v7) to use local version of `bulk` package as `v6` and `v7` use their own `bulk` packages. When implementing support for Elasticsearch v8 client, cadence-local `bulk` package will be used. 

re type-assert: done

Reply:
`os2` and `v8` are sufficiently similar that that'll be easy then?  could be reasonable, yea.  and not-migrating very old stuff seems fine too, we can consider that frozen unless forced to change.

any idea if we can remove `v6` any time soon?  7 came out in 2019.  
though surprisingly it seems they still offer support all the way back to _5.5_ which is ancient: https://www.elastic.co/support/matrix

Reply:
IMHO, it should be good to remove dependencies on old versions in general. It would be good idea to add support for `v8` and ind the same time, remove `v6` client